### PR TITLE
[location] Make get position async faster

### LINKED ISSUE #3433: Location.getCurrentPositionAsync() Takes 10+ seconds after upgrading to SDK 31
### Environment

```
  Expo CLI 2.6.14 environment info:
    System:
      OS: macOS 10.14
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 8.10.0 - ~/.nvm/versions/node/v8.10.0/bin/node
      Yarn: 1.13.0 - ~/.nvm/versions/node/v8.10.0/bin/yarn
      npm: 5.6.0 - ~/.nvm/versions/node/v8.10.0/bin/npm
      Watchman: 4.7.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.1 AI-173.4907809
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      @types/expo: ^31.0.1 => 31.0.7 
      @types/react: ^16.7.13 => 16.7.18 
      @types/react-native: ^0.57.18 => 0.57.27 
      expo: ^31.0.0 => 31.0.6 
      react: 16.5.0 => 16.5.0 
      react-native: https://github.com/expo/react-native/archive/sdk-31.0.1.tar.gz => 0.57.1 
      react-navigation: ^2.18.2 => 2.18.3 
    npmGlobalPackages:
      expo-cli: 2.6.14
```

This bug is happening on iOS in **standalone app only**

### Steps to Reproduce

I upgraded my app to SDK V31.  After noticing severly degraded performance related to location based services, I started profiling to see what was going on. The following consistently takes 10,100 ms to complete:

```js
import { Location } from 'expo';

async function getLocation() {
  const start = Date.now();
  await Location.getCurrentLocationAsync();
  console.log(Date.now() - start); // consistently around 10,100ms
}
```
This issue **only** happens in the standalone production version of my app. In development, there are no issues. I'd be happy to provide more information, but not sure what more to include given that this singular invocation is what's causing problems.

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Resolve #3433.

# How

Added two `MethodType` to getCurrentPosition. 
- `Slow` which work like before
- `Fast` which use `watchPositionAsync` to get current position.

It is faster on iOS because when we earlier used `requestLocation` method which obtain a location fix and it may take several seconds. 

On Android both method work in similar time on average but `Fast` method in some conditions will be working a little bit faster.   

# Test Plan

- test-suits passed on Android and iOS device.
- https://snack.expo.io/rJDocbIWr - no longer take 10 sec to get current position.

### CODE REVIEW

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: Beep

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: Beep

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: Beep

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: Please take a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function 
You do not have to explicit wrap function body into `Promise`, because the result from `async` function is implicitly wrapped in `Promise` 😉

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: In general follow these guidelines:
1. use `const` as default value binding
2. use `let` if binding should change (or primitive value should and would be changed)
3. do not use `var` in ES5 and above 😄 

Summing up -> change all `let`s to `const`s

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: same as above -> no binding changing -> use `const`

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: IMO it's strange initialisation, I would rather go with
```suggestion
        let bestPosition: LocationData | undefined;
```

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: Here just check whether `bestPosition` is falsy
```suggestion
          if (bestPosition) {
```

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: `current` means nothing in this context, I would suggest going with `currentTry` or sth like this

File: packages/expo-location/src/Location.ts
Comment (bbarthec) [inline packages/expo-location/src/Location.ts]: I believe it would be nice to have your changes extracted to separate function with some TSDoc saying that:
```ts
/**
 * hey, I'm going to find out local position by subscribing to position events and when 
 * conditions are satisfied I would return the results
 */
```

File: docs/pages/versions/unversioned/sdk/location.md
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/location.md]: Is it intentional to have this type not described?

File: docs/pages/versions/unversioned/sdk/location.md
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/location.md]: `Location.MethodType.MethodType.{Slow, Fast}`?
Shouldn't that be `Location.MethodType.{Slow, Fast}`?

**Review by bbarthec — CHANGES_REQUESTED** [review]
💯 Nice job! 😂 
I would rather wait for @tsapeta to approve this PR, so please to not merge without his approval. Please adjust the code to my comments or wait for his opinion.

Add `CHANGELOG.md` entry as well.

File: packages/expo-location/src/Location.ts
Comment (lukmccall) [reply packages/expo-location/src/Location.ts]: Ok, but I need to get resolve and reject function. Could I get them without wrapping in `Promise`?

**Review by tsapeta — COMMENTED** [review]
I'm not entirely convinced we want to have such options available for users. Basically, they can do this JS function on their side and that would fix their issue 🤔 Maybe we should just use event subscription on the iOS, native side of `getCurrentPositionAsync` and keep API simple? What do you think @sjchmiela?

File: CHANGELOG.md
Comment (bbarthec) [inline CHANGELOG.md]:
```suggestion
- Added second method to `Location.getCurrentPositionAsync`, which is able to calculate current position faster. ([#4953](https://github.com/expo/expo/pull/4953) by [@lukmccall](https://github.com/lukmccall))
```

**Review by bbarthec — APPROVED** [review]

**Comment by sjchmiela** [comment]
Do we still want to merge this PR? Why would anyone use `SLOW`, with `FAST` being so good?

**Pull request closed without merging**
